apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriberClient.cs

```diff
@@ -1520,78 +1568,70 @@ void RetryAcksAndExtends(bool? ack, AckError ackError, IEnumerable<TimedId> exte
-                        return;
-                    }
+                        // Only ids that exist in _retryableIds dictionary are added to retryableErrorIds from idsToRetry.
+                        // So, id will always exist in _retryableIds dictionary.
+                        var retryInfo = _retryableIds[id];
+                        var backoff = s_ackBackoff.NextBackoff(retryInfo.Backoff ?? TimeSpan.Zero);
+                        // We should retry only for specified timeout.
+                        if (_clock.GetCurrentDateTimeUtc() > retryInfo.FirstTimeOfFailureInUtc + TimeSpan.FromSeconds(timeoutInSeconds) - backoff)
+                        {
+                            // We are past the retry timeout. This id is not retry-able.
+                            // Remove this id from _retryableIds dictionary and continue.
+                            _ = _retryableIds.TryRemove(id, out _);
+                            continue;
+                        }
 
-                    var backoff = s_ackBackoff.NextBackoff(retryInfo.CurrentBackoff ?? TimeSpan.Zero);
-                    // We should retry only for specified timeout.
-                    // If current UTC time is greater than first UTC time of failure + timeout, we should not retry.
-                    if (DateTime.UtcNow > retryInfo.FirstTimeOfFailureInUtc + TimeSpan.FromSeconds(timeoutInSeconds) - backoff)
-                    {
-                        // We are past the retry timeout.
-                        return;
+                        // Update back-off of retryableErrorIds in _retryableIds dictionary.
+                        _retryableIds[id] = retryInfo.WithBackoff(backoff);
                     }
 
-                    // If we have ids to retry and we are not past the specified timeout, then retry.
-                    // Update backoff of retryableErrorIds in _retryableIds dictionary.
-                    UpdateAllRetryableErrorBackoff(retryableErrorIds, backoff);
-                    // Retry errors that are still temporary errors. They will be retried after backOff.
-                    Task delayTask = _scheduler.Delay(backoff, _softStopCts.Token);
-                    // TODO: Check, if the below expression can be simplified or optimized.
-                    Add(delayTask, new NextAction(false, hasAcksOrNacks
-                    ? () => { ackActionToRetry(retryableErrorIds); StartPush(); }
-                    : () => { extendActionToRetry(extendIds.Where(j => retryableErrorIds.Contains(j.Id))); StartPush(); }));
-                }
-
-                // Updates the backoff of the retryableErrorIds in _retryableIds dictionary.
-                void UpdateAllRetryableErrorBackoff(IEnumerable<string> ids, TimeSpan? backoff)
-                {
-                    foreach (var id in ids)
+                    // The ids that exist in both retryableErrorIds and _retryableIds now are the ones that need to be retried.
```

Comment:
I've got an idea for a refactoring that I think would simplify this - but it can wait until this PR is in.
(Basically it would avoid creating the the first set of `retryableErrorIds`, instead looping through `idsToRetry` and only creating a list of IDs we *genuinely* want to retry. I won't go into more detail now, but I should probably try to do it very soon after this is in, before I forget what I meant...)